sage_code/strong_uniform_bounds.py

```diff
@@ -0,0 +1,193 @@
+    License, or any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+    The authors can be reached at: barinder.s.banwait@gmail.com and
+    maarten@mderickx.nl.
+
+    ====================================================================
+
+"""
+
+from sage.all import Partitions, divisors, matrix, GF, lcm, gcd, prime_range, prod, Integer
+from itertools import product
+import logging
+from .common_utils import get_weil_polys, x
+from .pil_integers import collapse_tuple
+from .type_one_primes import cached_bad_formal_immersion_data
+
+logger = logging.getLogger(__name__)
+
+def is_non_increasing(t, part):
+
+    for j in range(len(t)-1):
+        if part[j] == part[j+1]:
+            if t[j] < t[j+1]:
+                return False
+    return True
+
+
+def get_beta_mats_with_pow(F, pow=1):
+
+    frob_polys = get_weil_polys(F)
+    output = [matrix.companion(f) ** pow for f in frob_polys]
+
+    # We also need to add \pm 1, \pm q
+
+    f1 = (x - 1) * (x + 1)
+    output.append(matrix.companion(f1))
+
+    q = F.cardinality()
+    fq = (x - q) * (x + q)
+    output.append(matrix.companion(fq))
+
+    return output
+
+def remove_duplicates(list_of_e_tuples, part):
+
+    output = []
+
+    for t in list_of_e_tuples:
+        if is_non_increasing(t, part):
+            output.append(t)
+    return output
+
+def splitting_types(d):
+    """Implements step 1 of Algorithm 5.1, the possible splitting types
+
+    Args:
+        d (int): degree of number field
+    """
+
+    types = []
+
+    for r in range(1,d+1):
+        parts = list(Partitions(d, length=r))        
+        for part in parts:
+            divisors_of_partition = [divisors(t) for t in part]
+            possible_es = list(product(*divisors_of_partition))
+            possible_es = remove_duplicates(possible_es, part)
+            for possible_e_vec in possible_es:
+                f_vec = [part[j]/possible_e_vec[j] for j in range(r)]
+                type_dict = {}
+                type_dict['r'] = r
+                type_dict['es'] = possible_e_vec
+                type_dict['fs'] = tuple(f_vec)
+                types.append(type_dict)
+    return types
+
+
+def bound_from_split_type(split_type, eps, q, known_mult_bound=0):
+    """This implements Algorithm 5.1 from the paper, and returns the integers
+       B_{eps,q} and B^*_{eps,q}.
+
+    Args:
+        d (int): degree of number field
+        eps (tuple): isogeny signature
+        q (int): auxiliary rational prime
+    """
+    frob_poly_mats = [get_beta_mats_with_pow(GF(q**f), pow=my_e) for my_e,f in zip(split_type['es'],split_type['fs'])]
```

Comment:
```suggestion
    frob_poly_mats = [get_beta_mats_with_pow(GF(q**f), pow=12*my_e) for my_e,f in zip(split_type['es'],split_type['fs'])]
```

I think there is a factor 12 missing here. Since in overleaf we have 12e_i as exponent of \beta_i
<img width="526" alt="Screenshot 2023-01-31 at 16 18 18" src="https://user-images.githubusercontent.com/773234/215800859-d1389e94-509b-450f-ba7b-7927ad57f4cd.png">

Reply:
oh yes quite right ... fixed in next commit